server/src/main/java/io/druid/server/QueryManager.java

```diff
@@ -53,16 +66,45 @@ public void registerQuery(Query query, final ListenableFuture future)
   {
     final String id = query.getId();
     queries.put(id, future);
+    if(queryDatasources.get(id) == null) {
+      queryDatasources.put(id, getDataSources(query));
+    }
     future.addListener(
         new Runnable()
         {
           @Override
           public void run()
           {
             queries.remove(id, future);
+            queryDatasources.remove(id);
           }
         },
         MoreExecutors.sameThreadExecutor()
     );
   }
+
+  public List<String> getQueryDatasources(final String queryId) {
+    return queryDatasources.get(queryId);
+  }
+
+  public List<String> getDataSources(final Query query) {
+    List<String> datasources = new ArrayList<>();
+    getDataSourcesHelper(query, datasources);
```

Comment:
can we just put the code from getDataSourcesHelper(..) here instead of creating that?